segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+                    int bytesToRead = Math.min(bytesRemaining, Math.toIntExact(chunkToReadFrom.getLength() - (currentOffset - startOffsetForCurrentChunk)));
+                    //assert bytesToRead != 0;
+
+                    if (currentOffset >= startOffsetForCurrentChunk + chunkToReadFrom.getLength()) {
+                        // The current chunk is over. Move to the next one.
+                        currentChunkName = chunkToReadFrom.getNextChunk();
+                        if (null != currentChunkName) {
+                            startOffsetForCurrentChunk += chunkToReadFrom.getLength();
+                            chunkToReadFrom = (ChunkMetadata) txn.get(currentChunkName);
+                            log.debug("{} read - reading from next chunk - segment={}, chunk={}", logPrefix, streamSegmentName, chunkToReadFrom);
+                        }
+                    } else {
+                        Preconditions.checkState(bytesToRead != 0, "bytesToRead is 0");
+                        // Read data from the chunk.
+                        ChunkHandle chunkHandle = chunkStorage.openRead(chunkToReadFrom.getName());
+                        int bytesRead = chunkStorage.read(chunkHandle, currentOffset - startOffsetForCurrentChunk, bytesToRead, buffer, currentBufferOffset);
+
+                        bytesRemaining -= bytesRead;
+                        currentOffset += bytesRead;
+                        currentBufferOffset += bytesRead;
+                        totalBytesRead += bytesRead;
+                    }
+                }
+
+                Duration elapsed = timer.getElapsed();
+                log.debug("{} read - segment={}, offset={}, bytesRead={}, latency={}.", logPrefix, handle.getSegmentName(), offset, totalBytesRead, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "read", traceId, handle, offset, totalBytesRead);
+                return totalBytesRead;
+            }
+        });
+    }
+
+    /**
+     * Gets current information about a StreamSegment.
+     *
+     * @param streamSegmentName The full name of the StreamSegment.
+     * @param timeout           Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will contain the information requested about the StreamSegment.
+     * If the operation failed, it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * </ul>
+     */
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        checkInitialized();
+        return execute(() -> {
+            long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+            Preconditions.checkNotNull(streamSegmentName, "streamSegmentName");
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
```

Comment:
I'm not sure we need a transaction here, can you justify?

Reply:
Transaction is also used for both read and writing of the data. This is so that while the transaction is still pending it needs to get the same value for key that it wrote earlier. 
For first read it will read from the table sgement/cache  and then this value will be cached in transaction. This avoids multiple calls to table segments to read same value.
 
I did not opt for creating separate mechanism for read only read. All transactions auto close. So if it is not committed then it is aborted.